File: dart/lib/src/sync/sync_manager.dart
import 'dart:async';
import 'package:monero_dart/src/transaction/models.dart';
import 'package:monero_dart/src/transaction/scanner.dart';

/// Sync state representing the current synchronization status
sealed class SyncState {}

/// Not started syncing
class SyncStateIdle extends SyncState {
  @override
  bool operator ==(Object other) => other is SyncStateIdle;
  
  @override
  int get hashCode => 0;
}

/// Currently syncing blocks
class SyncStateSyncing extends SyncState {
  final int currentHeight;
  final int targetHeight;
  final int blocksProcessed;
  final int startTime;
  
  SyncStateSyncing({
    required this.currentHeight,
    required this.targetHeight,
    required this.blocksProcessed,
    int? startTime,
  }) : startTime = startTime ?? DateTime.now().millisecondsSinceEpoch;
  
  double get progress => targetHeight > 0 ? currentHeight / targetHeight : 0.0;
  
  double get blocksPerSecond {
    final elapsed = (DateTime.now().millisecondsSinceEpoch - startTime) / 1000;
    return elapsed > 0 ? blocksProcessed / elapsed : 0.0;
  }
  
  double get estimatedSecondsRemaining {
    final remaining = targetHeight - currentHeight;
    return blocksPerSecond > 0 ? remaining / blocksPerSecond : double.infinity;
  }
  
  @override
  bool operator ==(Object other) =>
      other is SyncStateSyncing &&
      other.currentHeight == currentHeight &&
      other.targetHeight == targetHeight;
  
  @override
  int get hashCode => Object.hash(currentHeight, targetHeight);
}

/// Sync completed
class SyncStateSynced extends SyncState {
  final int height;
  
  SyncStateSynced(this.height);
  
  @override
  bool operator ==(Object other) =>
      other is SyncStateSynced && other.height == height;
  
  @override
  int get hashCode => height.hashCode;
}

/// Sync error occurred
class SyncStateError extends SyncState {
  final String message;
  final Object? cause;
  
  SyncStateError(this.message, [this.cause]);
  
  @override
  bool operator ==(Object other) =>
      other is SyncStateError && other.message == message;
  
  @override
  int get hashCode => message.hashCode;
}

/// Events emitted during synchronization
sealed class SyncEvent {}

/// New block processed
class BlockProcessedEvent extends SyncEvent {
  final int height;
  final List<int> hash;
  
  BlockProcessedEvent(this.height, this.hash);
}

/// New transaction found
class TransactionFoundEvent extends SyncEvent {
  final Transaction tx;
  
  TransactionFoundEvent(this.tx);
}

/// New output received
class OutputReceivedEvent extends SyncEvent {
  final OwnedOutput output;
  
  OutputReceivedEvent(this.output);
}

/// Output spent
class OutputSpentEvent extends SyncEvent {
  final List<int> keyImage;
  
  OutputSpentEvent(this.keyImage);
}

/// Chain reorganization detected
class ReorgDetectedEvent extends SyncEvent {
  final int fromHeight;
  final int toHeight;
  
  ReorgDetectedEvent(this.fromHeight, this.toHeight);
}

/// Sync progress update
class ProgressUpdateEvent extends SyncEvent {
  final int current;
  final int target;
  
  ProgressUpdateEvent(this.current, this.target);
}

/// Configuration for sync manager
class SyncConfig {
  /// Number of blocks to fetch in parallel
  final int batchSize;
  
  /// Delay between batches (ms)
  final int batchDelayMs;
  
  /// Number of confirmations required
  final int confirmations;
  
  /// Auto-restart on error
  final bool autoRetry;
  
  /// Maximum retry attempts
  final int maxRetries;
  
  /// Retry delay (ms)
  final int retryDelayMs;
  
  const SyncConfig({
    this.batchSize = 100,
    this.batchDelayMs = 100,
    this.confirmations = 10,
    this.autoRetry = true,
    this.maxRetries = 3,
    this.retryDelayMs = 5000,
  });
}

/// Block data from daemon
class BlockData {
  final int height;
  final List<int> hash;
  final int timestamp;
  final List<int> prevHash;
  final List<Transaction> transactions;
  
  BlockData({
    required this.height,
    required this.hash,
    required this.timestamp,
    required this.prevHash,
    required this.transactions,
  });
  
  @override
  bool operator ==(Object other) =>
      other is BlockData &&
      other.height == height &&
      _listEquals(other.hash, hash);
  
  @override
  int get hashCode => Object.hash(height, Object.hashAll(hash));
}

bool _listEquals<T>(List<T> a, List<T> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

/// Interface for block data provider (implemented by DaemonClient)
abstract class BlockProvider {
  Future<int> getHeight();
  Future<BlockData> getBlockByHeight(int height);
  Future<List<BlockData>> getBlocksByRange(int startHeight, int endHeight);
}

/// Interface for wallet storage
abstract class WalletStorage {
  Future<int> getLastSyncedHeight();
  Future<void> setLastSyncedHeight(int height);
  Future<List<int>?> getBlockHash(int height);
  Future<void> setBlockHash(int height, List<int> hash);
  Future<void> saveOutput(OwnedOutput output);
  Future<void> markOutputSpent(List<int> keyImage);
  Future<void> rollbackToHeight(int height);
}

/// Manages blockchain synchronization for a wallet.
/// 
/// Coordinates block fetching, transaction scanning, and state management.
class SyncManager {
  final ViewKeyScanner _scanner;
  final BlockProvider _blockProvider;
  final WalletStorage _storage;
  final SyncConfig _config;
  
  final _stateController = StreamController<SyncState>.broadcast();
  final _eventController = StreamController<SyncEvent>.broadcast();
  
  SyncState _currentState = SyncStateIdle();
  bool _isRunning = false;
  int _retryCount = 0;
  
  SyncManager({
    required ViewKeyScanner scanner,
    required BlockProvider blockProvider,
    required WalletStorage storage,
    SyncConfig config = const SyncConfig(),
  })  : _scanner = scanner,
        _blockProvider = blockProvider,
        _storage = storage,
        _config = config;
  
  /// Stream of sync state changes
  Stream<SyncState> get stateStream => _stateController.stream;
  
  /// Stream of sync events
  Stream<SyncEvent> get eventStream => _eventController.stream;
  
  /// Current sync state
  SyncState get state => _currentState;
  
  /// Whether currently syncing
  bool get isSyncing => _currentState is SyncStateSyncing;
  
  /// Start synchronization
  Future<void> start() async {
    if (_isRunning) return;
    _isRunning = true;
    
    try {
      await _sync();
    } catch (e) {
      await _handleError(e);
    }
  }
  
  /// Stop synchronization
  void stop() {
    _isRunning = false;
    _setState(SyncStateIdle());
  }
  
  /// Dispose resources
  void dispose() {
    stop();
    _stateController.close();
    _eventController.close();
  }
  
  void _setState(SyncState state) {
    _currentState = state;
    _stateController.add(state);
  }
  
  void _emitEvent(SyncEvent event) {
    _eventController.add(event);
  }
  
  Future<void> _sync() async {
    final startHeight = await _storage.getLastSyncedHeight() + 1;
    var targetHeight = await _blockProvider.getHeight();
    
    if (startHeight > targetHeight) {
      _setState(SyncStateSynced(targetHeight));
      return;
    }
    
    final startTime = DateTime.now().millisecondsSinceEpoch;
    var blocksProcessed = 0;
    var currentHeight = startHeight;
    
    _setState(SyncStateSyncing(
      currentHeight: currentHeight,
      targetHeight: targetHeight,
      blocksProcessed: blocksProcessed,
      startTime: startTime,
    ));
    
    while (currentHeight <= targetHeight && _isRunning) {
      // Check for reorg before processing
      if (currentHeight > 0) {
        final reorgHeight = await _checkForReorg(currentHeight - 1);
        if (reorgHeight != null) {
          currentHeight = await _handleReorg(reorgHeight);
          continue;
        }
      }
      
      // Fetch batch of blocks
      final endHeight = currentHeight + _config.batchSize - 1;
      final clampedEnd = endHeight > targetHeight ? targetHeight : endHeight;
      final blocks = await _blockProvider.getBlocksByRange(currentHeight, clampedEnd);
      
      // Process each block
      for (final block in blocks) {
        if (!_isRunning) break;
        
        await _processBlock(block);
        blocksProcessed++;
        currentHeight = block.height + 1;
        
        _setState(SyncStateSyncing(
          currentHeight: currentHeight,
          targetHeight: targetHeight,
          blocksProcessed: blocksProcessed,
          startTime: startTime,
        ));
        _emitEvent(ProgressUpdateEvent(currentHeight, targetHeight));
      }
      
      // Update target height (blockchain may have grown)
      final newTargetHeight = await _blockProvider.getHeight();
      if (newTargetHeight > targetHeight) {
        targetHeight = newTargetHeight;
        _setState(SyncStateSyncing(
          currentHeight: currentHeight,
          targetHeight: newTargetHeight,
          blocksProcessed: blocksProcessed,
          startTime: startTime,
        ));
      }
      
      // Small delay between batches
      if (_config.batchDelayMs > 0 && currentHeight <= targetHeight) {
        await Future<void>.delayed(Duration(milliseconds: _config.batchDelayMs));
      }
    }
    
    if (_isRunning) {
      _retryCount = 0;
      _setState(SyncStateSynced(currentHeight - 1));
    }
  }
  
  Future<void> _processBlock(BlockData block) async {
    // Save block hash for reorg detection
    await _storage.setBlockHash(block.height, block.hash);
    
    // Scan each transaction
    for (final tx in block.transactions) {
      final ownedOutputs = _scanner.scanTransaction(tx);
      
      if (ownedOutputs.isNotEmpty) {
        _emitEvent(TransactionFoundEvent(tx));
        
        for (final scanned in ownedOutputs) {
          if (scanned.isOwned) {
            final owned = OwnedOutput(
              txHash: tx.hash,
              outputIndex: scanned.output.index,
              globalIndex: scanned.output.globalIndex,
              amount: scanned.amount ?? 0,
              publicKey: scanned.output.target.key,
              blockHeight: block.height,
              timestamp: block.timestamp,
              subaddressMajor: scanned.subaddressIndex?.major ?? 0,
              subaddressMinor: scanned.subaddressIndex?.minor ?? 0,
            );
            await _storage.saveOutput(owned);
            _emitEvent(OutputReceivedEvent(owned));
          }
        }
      }
      
      // Check for spent outputs (key images in inputs)
      for (final input in tx.inputs) {
        await _storage.markOutputSpent(input.keyImage);
        _emitEvent(OutputSpentEvent(input.keyImage));
      }
    }
    
    // Update last synced height
    await _storage.setLastSyncedHeight(block.height);
    _emitEvent(BlockProcessedEvent(block.height, block.hash));
  }
  
  /// Check for blockchain reorganization
  /// Returns the height where reorg occurred, or null if no reorg
  Future<int?> _checkForReorg(int height) async {
    final storedHash = await _storage.getBlockHash(height);
    if (storedHash == null) return null;
    
    final block = await _blockProvider.getBlockByHeight(height);
    
    if (!_listEquals(block.hash, storedHash)) {
      // Reorg detected, find the fork point
      var checkHeight = height - 1;
      while (checkHeight > 0) {
        final stored = await _storage.getBlockHash(checkHeight);
        if (stored == null) break;
        final current = await _blockProvider.getBlockByHeight(checkHeight);
        if (_listEquals(current.hash, stored)) {
          return checkHeight + 1; // Fork point
        }
        checkHeight--;
      }
      return 0; // Complete reorg from genesis
    }
    
    return null;
  }
  
  /// Handle blockchain reorganization
  /// Returns the height to restart syncing from
  Future<int> _handleReorg(int reorgHeight) async {
    final currentHeight = await _storage.getLastSyncedHeight();
    _emitEvent(ReorgDetectedEvent(reorgHeight, currentHeight));
    
    // Rollback storage to before the reorg
    await _storage.rollbackToHeight(reorgHeight - 1);
    
    return reorgHeight;
  }
  
  Future<void> _handleError(Object error) async {
    _setState(SyncStateError(error.toString(), error));
    
    if (_config.autoRetry && _retryCount < _config.maxRetries) {
      _retryCount++;
      await Future<void>.delayed(Duration(milliseconds: _config.retryDelayMs));
      if (_isRunning) {
        await start();
      }
    }
  }
}

File: kmp/monero-core/src/commonMain/kotlin/io/monero/core/sync/SyncManager.kt
package io.monero.core.sync

import io.monero.core.scanner.ViewKeyScanner
import io.monero.core.transaction.OwnedOutput
import io.monero.core.transaction.Transaction
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

/**
 * Sync state representing the current synchronization status
 */
sealed class SyncState {
    /** Not started syncing */
    data object Idle : SyncState()
    
    /** Currently syncing blocks */
    data class Syncing(
        val currentHeight: Long,
        val targetHeight: Long,
        val blocksProcessed: Long,
        val startTime: Long = System.currentTimeMillis()
    ) : SyncState() {
        val progress: Float get() = if (targetHeight > 0) currentHeight.toFloat() / targetHeight else 0f
        val blocksPerSecond: Float get() {
            val elapsed = (System.currentTimeMillis() - startTime) / 1000f
            return if (elapsed > 0) blocksProcessed / elapsed else 0f
        }
        val estimatedSecondsRemaining: Float get() {
            val remaining = targetHeight - currentHeight
            return if (blocksPerSecond > 0) remaining / blocksPerSecond else Float.MAX_VALUE
        }
    }
    
    /** Sync completed */
    data class Synced(val height: Long) : SyncState()
    
    /** Sync error occurred */
    data class Error(val message: String, val cause: Throwable? = null) : SyncState()
}

/**
 * Events emitted during synchronization
 */
sealed class SyncEvent {
    /** New block processed */
    data class BlockProcessed(val height: Long, val hash: ByteArray) : SyncEvent()
    
    /** New transaction found */
    data class TransactionFound(val tx: Transaction) : SyncEvent()
    
    /** New output received */
    data class OutputReceived(val output: OwnedOutput) : SyncEvent()
    
    /** Output spent */
    data class OutputSpent(val keyImage: ByteArray) : SyncEvent()
    
    /** Chain reorganization detected */
    data class ReorgDetected(val fromHeight: Long, val toHeight: Long) : SyncEvent()
    
    /** Sync progress update */
    data class ProgressUpdate(val current: Long, val target: Long) : SyncEvent()
}

/**
 * Configuration for sync manager
 */
data class SyncConfig(
    /** Number of blocks to fetch in parallel */
    val batchSize: Int = 100,
    
    /** Delay between batches (ms) */
    val batchDelayMs: Long = 100,
    
    /** Number of confirmations required */
    val confirmations: Int = 10,
    
    /** Auto-restart on error */
    val autoRetry: Boolean = true,
    
    /** Maximum retry attempts */
    val maxRetries: Int = 3,
    
    /** Retry delay (ms) */
    val retryDelayMs: Long = 5000
)

/**
 * Block data from daemon
 */
data class BlockData(
    val height: Long,
    val hash: ByteArray,
    val timestamp: Long,
    val prevHash: ByteArray,
    val transactions: List<Transaction>
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false
        other as BlockData
        return height == other.height && hash.contentEquals(other.hash)
    }
    
    override fun hashCode(): Int = hash.contentHashCode()
}

/**
 * Interface for block data provider (implemented by DaemonClient)
 */
interface BlockProvider {
    suspend fun getHeight(): Long
    suspend fun getBlockByHeight(height: Long): BlockData
    suspend fun getBlocksByRange(startHeight: Long, endHeight: Long): List<BlockData>
}

/**
 * Interface for wallet storage
 */
interface WalletStorage {
    suspend fun getLastSyncedHeight(): Long
    suspend fun setLastSyncedHeight(height: Long)
    suspend fun getBlockHash(height: Long): ByteArray?
    suspend fun setBlockHash(height: Long, hash: ByteArray)
    suspend fun saveOutput(output: OwnedOutput)
    suspend fun markOutputSpent(keyImage: ByteArray)
    suspend fun rollbackToHeight(height: Long)
}

/**
 * Manages blockchain synchronization for a wallet.
 * 
 * Coordinates block fetching, transaction scanning, and state management.
 */
class SyncManager(
    private val scanner: ViewKeyScanner,
    private val blockProvider: BlockProvider,
    private val storage: WalletStorage,
    private val config: SyncConfig = SyncConfig(),
    private val dispatcher: CoroutineDispatcher = Dispatchers.Default
) {
    private val _state = MutableStateFlow<SyncState>(SyncState.Idle)
    val state: StateFlow<SyncState> = _state.asStateFlow()
    
    private val _events = MutableSharedFlow<SyncEvent>(replay = 0, extraBufferCapacity = 100)
    val events: SharedFlow<SyncEvent> = _events.asSharedFlow()
    
    private var syncJob: Job? = null
    private var retryCount = 0
    
    /**
     * Start synchronization from the last synced height
     */
    fun start(scope: CoroutineScope) {
        if (syncJob?.isActive == true) return
        
        syncJob = scope.launch(dispatcher) {
            try {
                sync()
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                handleError(e, scope)
            }
        }
    }
    
    /**
     * Stop synchronization
     */
    fun stop() {
        syncJob?.cancel()
        syncJob = null
        _state.value = SyncState.Idle
    }
    
    /**
     * Check if currently syncing
     */
    val isSyncing: Boolean get() = _state.value is SyncState.Syncing
    
    private suspend fun sync() {
        val startHeight = storage.getLastSyncedHeight() + 1
        val targetHeight = blockProvider.getHeight()
        
        if (startHeight > targetHeight) {
            _state.value = SyncState.Synced(targetHeight)
            return
        }
        
        val startTime = System.currentTimeMillis()
        var blocksProcessed = 0L
        var currentHeight = startHeight
        
        _state.value = SyncState.Syncing(currentHeight, targetHeight, blocksProcessed, startTime)
        
        while (currentHeight <= targetHeight) {
            // Check for reorg before processing
            if (currentHeight > 0) {
                val reorgHeight = checkForReorg(currentHeight - 1)
                if (reorgHeight != null) {
                    currentHeight = handleReorg(reorgHeight)
                    continue
                }
            }
            
            // Fetch batch of blocks
            val endHeight = minOf(currentHeight + config.batchSize - 1, targetHeight)
            val blocks = blockProvider.getBlocksByRange(currentHeight, endHeight)
            
            // Process each block
            for (block in blocks) {
                processBlock(block)
                blocksProcessed++
                currentHeight = block.height + 1
                
                _state.value = SyncState.Syncing(currentHeight, targetHeight, blocksProcessed, startTime)
                _events.emit(SyncEvent.ProgressUpdate(currentHeight, targetHeight))
            }
            
            // Update target height (blockchain may have grown)
            val newTargetHeight = blockProvider.getHeight()
            if (newTargetHeight > targetHeight) {
                _state.value = SyncState.Syncing(currentHeight, newTargetHeight, blocksProcessed, startTime)
            }
            
            // Small delay between batches to avoid overwhelming the node
            if (config.batchDelayMs > 0 && currentHeight <= targetHeight) {
                delay(config.batchDelayMs)
            }
        }
        
        retryCount = 0
        _state.value = SyncState.Synced(currentHeight - 1)
    }
    
    private suspend fun processBlock(block: BlockData) {
        // Save block hash for reorg detection
        storage.setBlockHash(block.height, block.hash)
        
        // Scan each transaction
        for (tx in block.transactions) {
            val ownedOutputs = scanner.scanTransaction(tx)
            
            if (ownedOutputs.isNotEmpty()) {
                _events.emit(SyncEvent.TransactionFound(tx))
                
                for (scanned in ownedOutputs) {
                    if (scanned.isOwned) {
                        val owned = OwnedOutput(
                            txHash = tx.hash,
                            outputIndex = scanned.output.index,
                            globalIndex = scanned.output.globalIndex,
                            amount = scanned.amount ?: 0L,
                            publicKey = scanned.output.target.key,
                            blockHeight = block.height,
                            timestamp = block.timestamp,
                            subaddressMajor = scanned.subaddressIndex?.first ?: 0,
                            subaddressMinor = scanned.subaddressIndex?.second ?: 0
                        )
                        storage.saveOutput(owned)
                        _events.emit(SyncEvent.OutputReceived(owned))
                    }
                }
            }
            
            // Check for spent outputs (key images in inputs)
            for (input in tx.inputs) {
                storage.markOutputSpent(input.keyImage)
                _events.emit(SyncEvent.OutputSpent(input.keyImage))
            }
        }
        
        // Update last synced height
        storage.setLastSyncedHeight(block.height)
        _events.emit(SyncEvent.BlockProcessed(block.height, block.hash))
    }
    
    /**
     * Check for blockchain reorganization
     * Returns the height where reorg occurred, or null if no reorg
     */
    private suspend fun checkForReorg(height: Long): Long? {
        val storedHash = storage.getBlockHash(height) ?: return null
        val block = blockProvider.getBlockByHeight(height)
        
        if (!block.hash.contentEquals(storedHash)) {
            // Reorg detected, find the fork point
            var checkHeight = height - 1
            while (checkHeight > 0) {
                val stored = storage.getBlockHash(checkHeight) ?: break
                val current = blockProvider.getBlockByHeight(checkHeight)
                if (current.hash.contentEquals(stored)) {
                    return checkHeight + 1 // Fork point
                }
                checkHeight--
            }
            return 0 // Complete reorg from genesis
        }
        
        return null
    }
    
    /**
     * Handle blockchain reorganization
     * Returns the height to restart syncing from
     */
    private suspend fun handleReorg(reorgHeight: Long): Long {
        val currentHeight = storage.getLastSyncedHeight()
        _events.emit(SyncEvent.ReorgDetected(reorgHeight, currentHeight))
        
        // Rollback storage to before the reorg
        storage.rollbackToHeight(reorgHeight - 1)
        
        return reorgHeight
    }
    
    private suspend fun handleError(e: Exception, scope: CoroutineScope) {
        _state.value = SyncState.Error(e.message ?: "Unknown error", e)
        
        if (config.autoRetry && retryCount < config.maxRetries) {
            retryCount++
            delay(config.retryDelayMs)
            start(scope)
        }
    }
}
